ch08performance/030misc-libs.ipynb.py

```diff
@@ -0,0 +1,385 @@
+# implementation! Why did this happen? Numba does not provide valuable performance
+# gains over pure Python or NumPy code for simple operations and small dataset.
+# The JITted function turned out to be slower than the non-JIT implementation
+# because of the compilation overhead. Let's try increasing the size of our
+# data and perform a non-NumPy list comprehension on the data.
+# 
+# The `jit` decorator with `nopython=True` is so widely used there exists an alias
+# decorator for the same - `@njit`
+
+# %%
+data = np.random.uniform(low=0.0, high=100.0, size=1_000_000)
+
+
+# %%
+def f(x):
+    return [math.sqrt(elem) for elem in x]
+
+@njit
+def jit_f(x):
+    return [math.sqrt(elem) for elem in x]
+
+# %%
+# %%timeit
+f(data)
+
+# %%
+# %%timeit
+jit_f(data)
+
+# %% [markdown]
+# That was way faster than the non-JIT function! But, the result was still slower
+# than the NumPy implementation. NumPy is still good for relatively simple
+# computations, but as the complexity increases, Numba functions start
+# outperforming NumPy implementations.
+
+# %% [markdown]
+# Let's go back to our plain Python mandelbrot code from the previous lessons and
+# JIT compile it -
+
+# %%
+@njit
+def mandel1(position, limit=50):
+    
+    value = position
+    
+    while abs(value) < 2:
+        limit -= 1        
+        value = value**2 + position
+        if limit < 0:
+            return 0
+        
+    return limit
+
+xmin = -1.5
+ymin = -1.0
+xmax = 0.5
+ymax = 1.0
+resolution = 300
+xstep = (xmax - xmin) / resolution
+ystep = (ymax - ymin) / resolution
+xs = [(xmin + (xmax - xmin) * i / resolution) for i in range(resolution)]
+ys = [(ymin + (ymax - ymin) * i / resolution) for i in range(resolution)]
+
+# %%
+# %%timeit
+data = [[mandel1(complex(x, y)) for x in xs] for y in ys]
+
+# %% [markdown]
+# The compiled code already beats our fastest NumPy implementation! It is not
+# necessary the compiled code will perform better than NumPy code, but it usually
+# gives performance gains for signifantly large computations. As always, it is
+# good to measure the perrformance to check if there are any gains.
+
+# %% [markdown]
+# Let's try JITting our NumPy code.
+
+# %%
+@njit
+def mandel_numpy(position,limit=50):
+    value = position
+    diverged_at_count = np.zeros(position.shape)
+    while limit > 0:
+        limit -= 1
+        value = value**2+position
+        diverging = value * np.conj(value) > 4
+        first_diverged_this_time = np.logical_and(diverging, diverged_at_count == 0)
+        diverged_at_count[first_diverged_this_time] = limit
+        value[diverging] = 2
+        
+    return diverged_at_count
+
+ymatrix, xmatrix = np.mgrid[ymin:ymax:ystep, xmin:xmax:xstep]
+values = xmatrix + 1j * ymatrix
+
+# %%
+# %%timeit
+mandel_numpy(values)
+
+
+# %% [markdown]
+# That does not work. The error might be solvable or it might just be out of Numba's
+# scope. Numba does not distinguish between plain Python
```

Comment:
It feels that we should give some insight into the error. ~~But I don't know exactly where it is...~~
[Edit]: After playing with it, I think I've managed to so by flatting the arrays (and taking only the real part on the comparison)
```python
@njit
def mandel_numpy(position,limit=50):
    value = position.flatten()
    diverged_at_count = np.zeros(position.shape).flatten()
    while limit > 0:
        limit -= 1
        value = value**2+position.flatten()
        diverging = (value * np.conj(value)).real > 4
        if limit == 49:  print(diverging.shape, diverged_at_count.shape)
        first_diverged_this_time = (np.logical_and(diverging, diverged_at_count == 0))
        diverging_at_count[first_diverged_this_time] = limit
        value[diverging] = 2

    return diverged_at_count.reshape(position.shape)

ymatrix, xmatrix = np.mgrid[ymin:ymax:ystep, xmin:xmax:xstep]
values = xmatrix + 1j * ymatrix
```

Reply:
Thank you for looking into this! 😄